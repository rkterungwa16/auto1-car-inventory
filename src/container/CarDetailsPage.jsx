import React, { Fragment } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import Header from '../components/Header'
import Footer from '../components/Footer'
import Button from '../components/Button'

import * as carsActions from '../actions/cars'

class HomePage extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      car: {},
      isGettingCar: false,
      addingCarError: null
    }
  }

  componentDidMount () {
    const {
      match
    } = this.props
    this.props.carsActions.getSingleCar(match.params.stockNumber)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.cars.car !== this.props.cars.car) {
      this.setState({
        car: nextProps.cars.car
      })
    }

    if (this.props.cars.isFetchingSingleCar !== nextProps.cars.isFetchingSingleCar) {
      this.setState({ isGettingCar: nextProps.cars.isFetchingSingleCar })
    }

    if (this.props.cars.addingCarError !== nextProps.cars.addingCarError) {
      this.setState({ addingCarError: nextProps.cars.addingCarError })
    }
  }

  /**
   * add particular car to favorites list
   * @param {Object} car Object containing car details
   * @return {*} none
   */
  addFavoriteCar (car) {
    this.props.carsActions.addFavoriteCar(car)
  }

  render () {
    const {
      isGettingCar
    } = this.state
    const {
      car
    } = this.state
    console.log('car details props', this.props.cars.addingCarError)
    if (isGettingCar) {
      return (
        <h1>Loading</h1>
      )
    }
    return (
      <Fragment>
        <Header />
        <div className='car-view'>
          <div className='car-details'>
            <h1
              className='car-details__text--xlarge car-details__text--margin'
            >
              {car.modelName}
            </h1>
            <h2
              className='car-details__text--large car-details__text--margin'
            >
              {`Stock #
              ${car.stockNumber} -
              ${car.mileage ? `${car.mileage.number} ${car.mileage.unit.toUpperCase()}` : 'none'}-
              ${car.fuelType} -
              ${car.color}`}
            </h2>
            <p className='car-details__text--small car-details__text--margin'>
            This car is currently available and can be delivered as soon as
            tomorrow morning. Please be aware that delivery times shown in
            this page are not definitive and may change due to bad weather
            conditions.
            </p>
          </div>
          <div className='save-car'>
            <div className='save-car__text'>
              <p className='save-car__text--small'>
              If you like this car, click the button and
              save it in your collection of favourite
              items.
              </p>
            </div>
            <div className='save-car__btn--wrapper'>
              <Button
                text='Save'
                handleClick={this.addFavoriteCar.bind(this, car)}
              />
            </div>
          </div>
        </div>
        <Footer />
      </Fragment>
    )
  }
}

const mapStateToProps = state => {
  return {
    cars: state.cars
  }
}

const mapDispatchToProps = dispatch => {
  return {
    dispatch,
    carsActions: bindActionCreators(carsActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(HomePage)
