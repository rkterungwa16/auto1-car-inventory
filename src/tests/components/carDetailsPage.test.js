import React from 'react'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import CarDetailsPage from '../../container/CarDetailsPage'

const state = {
  cars: { car: { modelName: 'i3' } }
}
const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const store = mockStore(state)

jest.genMockFromModule('uniqid')
Enzyme.configure({ adapter: new Adapter() })
const { shallow } = Enzyme

const carsActions = {
  getAllCars: jest.fn()
}

const wrapper = shallow(
  <CarDetailsPage
    store={store}
    carsActions={carsActions}
  />
)
test('should render CarDetailsPage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should render CarDetailsPage with alt data correctly', () => {
  wrapper.setProps({
    cars: { car: { modelName: 'car' } }
  })
  expect(wrapper).toMatchSnapshot()
})
