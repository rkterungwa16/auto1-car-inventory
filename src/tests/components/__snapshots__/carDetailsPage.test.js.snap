// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CarDetailsPage correctly 1`] = `
<HomePage
  cars={
    Object {
      "car": Object {
        "modelName": "i3",
      },
    }
  }
  carsActions={
    Object {
      "addFavoriteCar": [Function],
      "addFavoriteCarFailure": [Function],
      "addFavoriteCarRequest": [Function],
      "addFavoriteCarSuccess": [Function],
      "getAllCars": [Function],
      "getCarsFailure": [Function],
      "getCarsRequest": [Function],
      "getCarsSuccess": [Function],
      "getSingleCar": [Function],
      "getSingleCarFailure": [Function],
      "getSingleCarRequest": [Function],
      "getSingleCarSuccess": [Function],
      "removeFavoriteCar": [Function],
      "removeFavoriteCarFailure": [Function],
      "removeFavoriteCarRequest": [Function],
      "removeFavoriteCarSuccess": [Function],
    }
  }
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`should render CarDetailsPage with alt data correctly 1`] = `
<HomePage
  cars={
    Object {
      "car": Object {
        "modelName": "i3",
      },
    }
  }
  carsActions={
    Object {
      "addFavoriteCar": [Function],
      "addFavoriteCarFailure": [Function],
      "addFavoriteCarRequest": [Function],
      "addFavoriteCarSuccess": [Function],
      "getAllCars": [Function],
      "getCarsFailure": [Function],
      "getCarsRequest": [Function],
      "getCarsSuccess": [Function],
      "getSingleCar": [Function],
      "getSingleCarFailure": [Function],
      "getSingleCarRequest": [Function],
      "getSingleCarSuccess": [Function],
      "removeFavoriteCar": [Function],
      "removeFavoriteCarFailure": [Function],
      "removeFavoriteCarRequest": [Function],
      "removeFavoriteCarSuccess": [Function],
    }
  }
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
